fix(compiler): incorrectly inferring namespace for HTML nodes inside SVG

crisbeto opened the pull request:
The HTML parser gets an element's namespace either from the tag name (e.g. `<svg:rect>`) or from its parent element `<svg><rect></svg>`) which breaks down when an element is inside of an SVG `foreignElement`, because foreign elements allow nodes from a different namespace to be inserted into an SVG.

These changes add another flag to the tag definitions which tells child nodes whether to try to inherit their namespaces from their parents. It also adds a definition for `foreignObject` with the new flag, allowing elements placed inside it to infer their namespaces instead.

Fixes #37218.

alxhub commented on packages/compiler/src/ml_parser/tags.ts:
```diff
@@ -52,6 +52,11 @@ export function isNgTemplate(tagName: string): boolean {
   return splitNsName(tagName)[1] === 'ng-template';
 }
 
+export function isSvgForeignObject(tagName: string): boolean {
```
Is there a way to put this knowledge on the tag definition for `svg:foreignObject` rather than special casing it in the parser?

crisbeto replied:
Done.

AndrewKushnir commented on packages/compiler/src/ml_parser/html_tags.ts:
```diff
@@ -88,6 +92,8 @@ export function getHtmlTagDefinition(tagName: string): HtmlTagDefinition {
       'col': new HtmlTagDefinition({isVoid: true}),
       'svg': new HtmlTagDefinition({implicitNamespacePrefix: 'svg'}),
+      'foreignObject': new HtmlTagDefinition(
+          {implicitNamespacePrefix: 'svg', preventNamespaceInheritance: true}),
```
I think it might make sense to add a quick comment here to explain why we want to avoid namespace inheritance in case of `foreignObject`.

crisbeto replied:
Done.

AndrewKushnir commented on packages/compiler/src/ml_parser/html_tags.ts:
```diff
@@ -111,5 +117,9 @@ export function getHtmlTagDefinition(tagName: string): HtmlTagDefinition {
-  return TAG_DEFINITIONS[tagName.toLowerCase()] || _DEFAULT_TAG_DEFINITION;
+  // The fallback lowercase-based lookup is required for ViewEngine which seems to be passing the
+  // tag names in upper case in some cases. We can remove it once ViewEngine is gone since Ivy
+  // preserves the tag name casing which is important for SVG nodes like `foreignObject`.
```
Quick question: would the `tagName` contain preserved casing if uppercase tag names are used in a template like this:
```
<DIV>Some text</DIV>
```
If yes (i.e. `tagName` is `DIV`), we'd need the `TAG_DEFINITIONS[tagName.toLowerCase()]` logic even after View Engine is removed (and the first condition would not be needed I guess).

crisbeto replied:
Good point, I tried it out and that is indeed the case. I've reworded the comment so we don't remove the extra lookup in the future.

AndrewKushnir replied:
Thanks for updating the comment @crisbeto. I'm wondering if we actually need a case-sensitive lookup. Can we always lowercase the tag name and check it against the set of tags? For example, if the `<ForeignObject>` tag works correctly in HTML, we would not match it at all using the current logic.

crisbeto replied:
As far as I know, `foreignObject` can't be written as `ForeignObject`. There are a handful of SVG elements attributes like these that have to be written in camel case. We should be able to get away with the extra lookup here since it's constant time and it's very likely to match on the first try anyway since most people write the node names in lower case.

AndrewKushnir replied:
Got it, thanks! 👍

AndrewKushnir reviewed (commented):
Thanks for the fix @crisbeto, I just came across this PR and left a couple comments.

alxhub commented on packages/compiler/src/ml_parser/html_tags.ts:
```diff
@@ -111,5 +119,8 @@ export function getHtmlTagDefinition(tagName: string): HtmlTagDefinition {
-  return TAG_DEFINITIONS[tagName.toLowerCase()] || _DEFAULT_TAG_DEFINITION;
+  // We have to make both a case-sensitive and a case-insesitive lookup, because
+  // HTML tag names are case insensitive, whereas some SVG tags are case sensitive.
+  return TAG_DEFINITIONS[tagName] || TAG_DEFINITIONS[tagName.toLowerCase()] ||
```
I know the old code uses `||`, but this is a great opportunity to switch to `??` instead of `||` :)

alxhub commented on packages/compiler/src/ml_parser/html_tags.ts:
```diff
@@ -88,6 +92,10 @@ export function getHtmlTagDefinition(tagName: string): HtmlTagDefinition {
+      'foreignObject': new HtmlTagDefinition(
+          // We want to prevent children of foreignObject from inheriting its namespace, because
+          // the point of the element is to allow nodes from other namespaces to be inserted.
+          {implicitNamespacePrefix: 'svg', preventNamespaceInheritance: true}),
```
Can you add a bit to the comment above explaining why this needs an `implicitNamespacePrefix`? It's because the closing tag inherits the namespace of the opening tag, right? I would kind of prefer we fix that directly instead of using this "hack", but I think it's probably okay.

alxhub commented:
[Presubmit](https://test.corp.google.com/ui#id=OCL:328382642:BASE:328382671:1598384275229:dff20f26)

crisbeto committed:
fix(compiler): incorrectly inferring namespace for HTML nodes inside SVG

The HTML parser gets an element's namespace either from the tag name
(e.g. `<svg:rect>`) or from its parent element `<svg><rect></svg>`) which
breaks down when an element is inside of an SVG `foreignElement`,
because foreign elements allow nodes from a different namespace to be
inserted into an SVG.

These changes add another flag to the tag definitions which tells child
nodes whether to try to inherit their namespaces from their parents.
It also adds a definition for `foreignObject` with the new flag,
allowing elements placed inside it to infer their namespaces instead.

Fixes #37218.
